updater/servicecontrol_windows.go

```
@@ -110,28 +141,21 @@ func StopService(serviceName string) error {
 		return err
 	}
 
-	var status svc.Status
-	var retries int = 3
-	var sleep time.Duration = 5 * time.Second
+	var retries int = 90
 
 	for i := 0; i < retries; i++ {
-		running, _ := IsServiceRunning(serviceName)
+		time.Sleep(1 * time.Second)
+		stopped, err := IsServiceStopped(serviceName)
 
-		if running != true {
+		if err == nil && stopped {
 			// Returns nil as service is no longer running
 			return nil
 		}
-
-		// set status for exit condition
-		status, _ = s.Query() // what is the status of hupdater
-
-		// Wait for service to stop
-		time.Sleep(sleep)
-		sleep *= 5
 	}
 
+	status, _ := s.Query()
 	if status.State != svc.Stopped {
-		err = fmt.Errorf("'%s' did not stop; status: %+v", serviceName, status)
+		err = fmt.Errorf("'%s' did not stop in time; status: %+v", serviceName, status)
 		return err
 	}
```

If we get to the end of the loop we pretty much know that the service is not stopped so we can just return an error here.

Just as a by-the-way, what is the difference between the `s.Query` used here and the `s.Control(svc.Interrogate)` used above?

---

It looks like `s.Query` will call `QueryServiceStatusEx` internally and `s.Control(svc.Interrogate)` will actually send an event to the service control handler and wait for a response. I'm going to modify this to use query only.